File: apps/website/src/hooks/tags.ts
"use client";

import { useAuth } from "@clerk/nextjs";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { useActiveWorkspace } from "@/components/workspaces";
import { useGraphQL } from "@/lib/client";
import * as tagsService from "@/lib/service/tags";
import type {
  ReorderTagCategoryItem,
  Tag,
  TagCategory,
} from "@/lib/types/tags";
import {
  optimisticCreate,
  optimisticList,
  optimisticRemove,
  optimisticUpdate,
  tempId,
} from "./optimistic";

// ---------------------------------------------------------------------------
// Query keys
// ---------------------------------------------------------------------------

const TAGS_KEY = ["tags"] as const;

const categoriesKey = () => [...TAGS_KEY, "categories"] as const;

/** Narrow prefix covering every per-category tag list; deliberately excludes categories. */
const tagsListKey = [...TAGS_KEY, "list"] as const;

const tagsKey = (categoryId?: string) =>
  [...TAGS_KEY, "list", categoryId ?? null] as const;

const now = () => new Date().toISOString();

// ---------------------------------------------------------------------------
// Query hooks
// ---------------------------------------------------------------------------

export function useTagCategories() {
  const { isLoaded, isSignedIn } = useAuth();
  const fetcher = useGraphQL();
  const workspace = useActiveWorkspace();

  return useQuery<TagCategory[]>({
    queryKey: [...categoriesKey(), workspace?.id ?? null],
    queryFn: () => tagsService.fetchTagCategories(fetcher, workspace!.id),
    enabled: isLoaded && isSignedIn && !!workspace,
  });
}

export function useTags(categoryId?: string) {
  const { isLoaded, isSignedIn } = useAuth();
  const fetcher = useGraphQL();
  const workspace = useActiveWorkspace();

  return useQuery<Tag[]>({
    queryKey: [...tagsKey(categoryId), workspace?.id ?? null],
    queryFn: () => tagsService.fetchTags(fetcher, workspace!.id, categoryId),
    enabled: isLoaded && isSignedIn && !!workspace,
  });
}

/** Fetch all tags (no category filter). */
export function useAllTags() {
  return useTags();
}

// ---------------------------------------------------------------------------
// Tag category mutation hooks
// ---------------------------------------------------------------------------

export function useCreateTagCategory() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();
  const workspace = useActiveWorkspace();

  return useMutation({
    mutationFn: ({ name, color }: { name: string; color?: string | null }) =>
      workspace
        ? tagsService.createTagCategory(fetcher, workspace.id, name, color)
        : Promise.reject(new Error("Select a workspace first")),
    ...optimisticCreate<{ name: string; color?: string | null }, TagCategory>(
      queryClient,
      categoriesKey(),
      ({ name, color }) => ({
        id: tempId(),
        userId: "",
        workspaceId: workspace?.id ?? "",
        name,
        role: null,
        color: color ?? null,
        sortOrder: 0,
        createdAt: now(),
        updatedAt: now(),
      }),
    ),
  });
}

export function useRenameTagCategory() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ id, name }: { id: string; name: string }) =>
      tagsService.renameTagCategory(fetcher, id, name),
    ...optimisticUpdate<{ id: string; name: string }, TagCategory>(
      queryClient,
      categoriesKey(),
      (vars) => vars.id,
      (entity, { name }) => ({ ...entity, name }),
    ),
  });
}

export function useSetTagCategoryColor() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ id, color }: { id: string; color: string | null }) =>
      tagsService.setTagCategoryColor(fetcher, id, color),
    ...optimisticUpdate<{ id: string; color: string | null }, TagCategory>(
      queryClient,
      categoriesKey(),
      (vars) => vars.id,
      (entity, { color }) => ({ ...entity, color }),
    ),
  });
}

export function useReorderTagCategories() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (order: ReorderTagCategoryItem[]) =>
      tagsService.reorderTagCategories(fetcher, order),
    ...optimisticList<ReorderTagCategoryItem[], TagCategory>(
      queryClient,
      categoriesKey(),
      (list, order) => {
        const rank = new Map(order.map((o, i) => [o.id, i]));
        return [...list].sort(
          (a, b) =>
            (rank.get(a.id) ?? a.sortOrder) - (rank.get(b.id) ?? b.sortOrder),
        );
      },
    ),
  });
}

export function useDeleteTagCategory() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (id: string) => tagsService.deleteTagCategory(fetcher, id),
    // Deleting a category cascades to its tags, so reconcile the whole tags tree.
    ...optimisticRemove<string>(
      queryClient,
      categoriesKey(),
      (id) => id,
      TAGS_KEY,
    ),
  });
}

// ---------------------------------------------------------------------------
// Tag mutation hooks
// ---------------------------------------------------------------------------

export function useCreateTag() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();
  const workspace = useActiveWorkspace();

  type CreateTagVars = {
    categoryId: string;
    name: string;
    color?: string | null;
  };
  return useMutation({
    mutationFn: ({ categoryId, name, color }: CreateTagVars) =>
      workspace
        ? tagsService.createTag(fetcher, workspace.id, categoryId, name, color)
        : Promise.reject(new Error("Select a workspace first")),
    ...optimisticCreate<CreateTagVars, Tag>(
      queryClient,
      tagsListKey,
      ({ categoryId, name, color }) => ({
        id: tempId(),
        userId: "",
        workspaceId: workspace?.id ?? "",
        categoryId,
        name,
        color: color ?? null,
        createdAt: now(),
        updatedAt: now(),
      }),
      TAGS_KEY,
    ),
  });
}

export function useRenameTag() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ id, name }: { id: string; name: string }) =>
      tagsService.renameTag(fetcher, id, name),
    ...optimisticUpdate<{ id: string; name: string }, Tag>(
      queryClient,
      tagsListKey,
      (vars) => vars.id,
      (entity, { name }) => ({ ...entity, name }),
      TAGS_KEY,
    ),
  });
}

export function useSetTagColor() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ id, color }: { id: string; color: string | null }) =>
      tagsService.setTagColor(fetcher, id, color),
    ...optimisticUpdate<{ id: string; color: string | null }, Tag>(
      queryClient,
      tagsListKey,
      (vars) => vars.id,
      (entity, { color }) => ({ ...entity, color }),
      TAGS_KEY,
    ),
  });
}

export function useDeleteTag() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (id: string) => tagsService.deleteTag(fetcher, id),
    ...optimisticRemove<string>(queryClient, tagsListKey, (id) => id, TAGS_KEY),
  });
}

export function useMergeTags() {
  const fetcher = useGraphQL();
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ fromId, intoId }: { fromId: string; intoId: string }) =>
      tagsService.mergeTags(fetcher, fromId, intoId),
    // Merge folds one tag into another and re-points trade links; drop the source now,
    // let the settle refetch bring the reconciled trade-tag state.
    ...optimisticRemove<{ fromId: string; intoId: string }>(
      queryClient,
      tagsListKey,
      (vars) => vars.fromId,
      TAGS_KEY,
    ),
  });
}

File: backend/src/service/read_service/tags.rs
use anyhow::Result;

use crate::service::db::client::UserDb;
use crate::service::db::schema::tables::tags_table::{self, Tag, TagCategory};

// ---------------------------------------------------------------------------
// Categories
// ---------------------------------------------------------------------------

pub async fn list_categories(user_db: &UserDb, workspace_id: &str) -> Result<Vec<TagCategory>> {
    tags_table::list_categories(user_db.pool(), user_db.user_id(), workspace_id).await
}

pub async fn get_category(user_db: &UserDb, id: &str) -> Result<Option<TagCategory>> {
    tags_table::find_category(user_db.pool(), user_db.user_id(), id).await
}

pub async fn create_category(
    user_db: &UserDb,
    workspace_id: &str,
    name: &str,
    color: Option<&str>,
) -> Result<TagCategory> {
    tags_table::create_category(user_db.pool(), user_db.user_id(), workspace_id, name, color).await
}

pub async fn rename_category(user_db: &UserDb, id: &str, name: &str) -> Result<TagCategory> {
    tags_table::rename_category(user_db.pool(), user_db.user_id(), id, name).await
}

pub async fn set_category_color(
    user_db: &UserDb,
    id: &str,
    color: Option<&str>,
) -> Result<TagCategory> {
    tags_table::set_category_color(user_db.pool(), user_db.user_id(), id, color).await
}

pub async fn reorder_categories(user_db: &UserDb, order: &[(String, i64)]) -> Result<()> {
    tags_table::reorder_categories(user_db.pool(), user_db.user_id(), order).await
}

pub async fn delete_category(user_db: &UserDb, id: &str) -> Result<bool> {
    tags_table::delete_category(user_db.pool(), user_db.user_id(), id).await
}

// ---------------------------------------------------------------------------
// Tags
// ---------------------------------------------------------------------------

pub async fn list_tags(
    user_db: &UserDb,
    workspace_id: &str,
    category_id: Option<&str>,
) -> Result<Vec<Tag>> {
    tags_table::list_tags(user_db.pool(), user_db.user_id(), workspace_id, category_id).await
}

pub async fn get_tag(user_db: &UserDb, id: &str) -> Result<Option<Tag>> {
    tags_table::find_tag(user_db.pool(), user_db.user_id(), id).await
}

pub async fn create_tag(
    user_db: &UserDb,
    workspace_id: &str,
    category_id: &str,
    name: &str,
    color: Option<&str>,
) -> Result<Tag> {
    tags_table::create_tag(
        user_db.pool(),
        user_db.user_id(),
        workspace_id,
        category_id,
        name,
        color,
    )
    .await
}

pub async fn rename_tag(user_db: &UserDb, id: &str, name: &str) -> Result<Tag> {
    tags_table::rename_tag(user_db.pool(), user_db.user_id(), id, name).await
}

pub async fn set_tag_color(user_db: &UserDb, id: &str, color: Option<&str>) -> Result<Tag> {
    tags_table::set_tag_color(user_db.pool(), user_db.user_id(), id, color).await
}

pub async fn delete_tag(user_db: &UserDb, id: &str) -> Result<bool> {
    tags_table::delete_tag(user_db.pool(), user_db.user_id(), id).await
}

pub async fn merge_tags(user_db: &UserDb, from_id: &str, into_id: &str) -> Result<()> {
    tags_table::merge_tags(user_db.pool(), user_db.user_id(), from_id, into_id).await
}

// ---------------------------------------------------------------------------
// Trade <-> tag links
// ---------------------------------------------------------------------------

pub async fn set_trade_tags(
    user_db: &UserDb,
    journal_entry_id: &str,
    tag_ids: &[String],
) -> Result<()> {
    tags_table::set_trade_tags(user_db.pool(), user_db.user_id(), journal_entry_id, tag_ids).await
}

pub async fn tags_for_trade(user_db: &UserDb, journal_entry_id: &str) -> Result<Vec<Tag>> {
    tags_table::tags_for_trade(user_db.pool(), user_db.user_id(), journal_entry_id).await
}
